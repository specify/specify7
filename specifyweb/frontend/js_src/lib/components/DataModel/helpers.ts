import { f } from '../../utils/functools';
import type { RA, ValueOf } from '../../utils/types';
import { isTreeResource } from '../InitialContext/treeRanks';
import { relationshipIsToMany } from '../WbPlanView/mappingHelpers';
import type {
  AnySchema,
  AnyTree,
  SerializedResource,
  TableFields,
} from './helperTypes';
import type { SpecifyResource } from './legacyTypes';
import type { LiteralField, Relationship } from './specifyField';
import type { SpecifyTable } from './specifyTable';
import type { Tables } from './types';
/**
 * Lookups for operaters and relationships in Queries on the backend are
 * separated by `__`
 * The api supports the same syntax in query paramters.
 * For example `/api/specify/collectionobject/?collection__discipline__name__startswith="Invert"&catalognumber__gt=000000100`
 * fetches all collectionobjects in disciplines which names start with
 * "Invert" with catalognumbers greater than 100
 */
export const djangoLookupSeparator = '__';

export const backboneFieldSeparator = '.';

/**
 * Formats a relationship lookup which is used in a query passed to the backend.
 * For example `formatRelationshipPath('collection', 'discipline', 'division')`
 * becomes `'collection__discipline__division'`
 */
export const formatRelationshipPath = (...fields: RA<string>): string =>
  fields.join(djangoLookupSeparator);

const weekDayMap = {
  Sunday: 1,
  Monday: 2,
  Tuesday: 3,
  Wednesday: 4,
  Thursday: 5,
  Friday: 6,
  Saturday: 7,
};

/**
 * Use this to construct a query using a lookup for Django.
 * Returns an object which can be used as a filter when fetched from the backend.
 * Example: backendFilter('number1').isIn([1, 2, 3]) is the equivalent
 * of {number1__in: [1, 2, 3].join(',')}
 *
 * See the Django docs at:
 * https://docs.djangoproject.com/en/3.2/ref/models/querysets/#field-lookups
 */
export const backendFilter = (field: string) => ({
  equals: (value: number | string) => ({
    [[field, 'exact'].join(djangoLookupSeparator)]: value,
  }),
  contains: (value: string) => ({
    [[field, 'contains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveContains: (value: string) => ({
    [[field, 'icontains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveStartsWith: (value: string) => ({
    [[field, 'istartswith'].join(djangoLookupSeparator)]: value,
  }),
  startsWith: (value: string) => ({
    [[field, 'startswith'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveEndsWith: (value: string) => ({
    [[field, 'iendswith'].join(djangoLookupSeparator)]: value,
  }),
  endsWith: (value: string) => ({
    [[field, 'endswith'].join(djangoLookupSeparator)]: value,
  }),
  isIn: (value: RA<number | string>) => ({
    [[field, 'in'].join(djangoLookupSeparator)]: value.join(','),
  }),
  isNull: (value: 'false' | 'true' = 'true') => ({
    [[field, 'isnull'].join(djangoLookupSeparator)]: value,
  }),
  greaterThan: (value: number) => ({
    [[field, 'gt'].join(djangoLookupSeparator)]: value,
  }),
  greaterThanOrEqualTo: (value: number) => ({
    [[field, 'gte'].join(djangoLookupSeparator)]: value,
  }),
  lessThan: (value: number) => ({
    [[field, 'lt'].join(djangoLookupSeparator)]: value,
  }),
  lessThanOrEqualTo: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  matchesRegex: (value: string) => ({
    [[field, 'regex'].join(djangoLookupSeparator)]: value,
  }),

  dayEquals: (value: number) => ({
    [[field, 'day'].join(djangoLookupSeparator)]: value,
  }),
  monthEquals: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  yearEquals: (value: number) => ({
    [[field, 'year'].join(djangoLookupSeparator)]: value,
  }),
  weekEquals: (value: number) => ({
    [[field, 'week'].join(djangoLookupSeparator)]: value,
  }),
  weekDayEquals: (value: keyof typeof weekDayMap) => ({
    [[field, 'week_day'].join(djangoLookupSeparator)]: weekDayMap[value],
  }),
});

export const isResourceOfType = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
  // @ts-expect-error
): resource is SpecifyResource<Tables[TABLE_NAME]> =>
  resource.specifyTable.name === tableName;

export const toTable = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  resource.specifyTable.name === tableName
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

export const toResource = <TABLE_NAME extends keyof Tables>(
  resource: SerializedResource<AnySchema>,
  tableName: TABLE_NAME
): SerializedResource<Tables[TABLE_NAME]> | undefined =>
  resource._tableName === tableName
    ? (resource as SerializedResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * The table.field has a very broad type to reduce type conflicts in components
 * that deal with generic schemas (accept AnySchema or a SCHEMA extends AnySchema)
 */
export const getField = <
  SCHEMA extends ValueOf<Tables>,
  FIELD extends TableFields<SCHEMA>
>(
  table: SpecifyTable<SCHEMA>,
  name: FIELD
): FIELD extends keyof SCHEMA['fields'] ? LiteralField : Relationship =>
  table.field[name] as FIELD extends keyof SCHEMA['fields']
    ? LiteralField
    : Relationship;

export const toTreeTable = (
  resource: SpecifyResource<AnySchema>
): SpecifyResource<AnyTree> | undefined =>
  isTreeResource(resource) ? resource : undefined;

export const toTables = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableNames: RA<TABLE_NAME>
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  f.includes(tableNames, resource.specifyTable.name)
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * Example usage:
 * resource: Collector
 * fields: agent -> lastName
 * Would return [agent, lastName] if agent exists
 *
 */
export async function fetchDistantRelated(
  resource: SpecifyResource<AnySchema>,
  fields: RA<LiteralField | Relationship> | undefined
): Promise<
  | {
      readonly resource: SpecifyResource<AnySchema> | undefined;
      readonly field: LiteralField | Relationship | undefined;
    }
  | undefined
> {
  if (
    Array.isArray(fields) &&
    fields.some(
      (field) =>
        field.isRelationship &&
        relationshipIsToMany(field) &&
        field !== fields.at(-1)
    )
  ) {
    console.error(
      'Can not index inside of a -to-many relationship. Use an aggregator instead'
    );
    return undefined;
  }

  const related =
    fields === undefined || fields.length === 0
      ? resource
      : fields.length === 1
      ? await resource.fetch()
      : await resource.rgetPromise(
          fields
            .slice(0, -1)
            .map(({ name }) => name)
            .join(backboneFieldSeparator)
        );

  const field = fields?.at(-1);
  const relatedResource = related ?? undefined;
  return relatedResource === undefined && field === undefined
    ? undefined
    : {
        resource: relatedResource,
        field,
      };
}
