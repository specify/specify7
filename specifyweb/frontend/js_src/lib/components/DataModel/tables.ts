/**
 * Fetch Specify DataModel JSON and reify it into the SpecifyTable and
 * SpecifyField objects
 */

import type { LocalizedString } from 'typesafe-i18n';

import { f } from '../../utils/functools';
import type { IR, RA, RR } from '../../utils/types';
import {
  defined,
  overwriteReadOnly,
  setDevelopmentGlobal,
} from '../../utils/types';
import { sortFunction, split } from '../../utils/utils';
import { error } from '../Errors/assert';
import { load } from '../InitialContext';
import { isTreeTable } from '../InitialContext/treeRanks';
import { formatUrl } from '../Router/queryString';
import type { AnySchema, AnyTree } from './helperTypes';
import { schemaExtras } from './schemaExtras';
import { LiteralField, Relationship } from './specifyField';
import { type TableDefinition, SpecifyTable } from './specifyTable';
import type { Agent, Tables } from './types';

export type SchemaLocalization = {
  readonly name: LocalizedString | null;
  readonly desc: LocalizedString | null;
  readonly format: string | null;
  readonly aggregator: string | null;
  readonly ishidden: boolean;
  readonly items: IR<{
    readonly name: LocalizedString | null;
    readonly desc: LocalizedString | null;
    readonly format: string | null;
    readonly picklistname: string | null;
    readonly weblinkname: string | null;
    readonly isrequired: boolean;
    readonly ishidden: boolean;
  }>;
};

const processFields = <FIELD_TYPE extends LiteralField | Relationship>(
  fields: RA<FIELD_TYPE>,
  frontEndFields: RA<FIELD_TYPE>
): RA<FIELD_TYPE> =>
  [
    ...fields,
    ...frontEndFields.map((field) => {
      field.overrides.isReadOnly = true;
      field.isVirtual = true;
      return field;
    }),
    /*
     * The sort order defined here affects the order of fields in the
     * WbPlanView, Query builder, Schema Config and all other places
     */
  ].sort(sortFunction(({ label = '' }) => label));

let schemaLocalization: IR<SchemaLocalization> = undefined!;
const fetchSchemaLocalization = f.store(async () =>
  import('../Preferences/userPreferences').then(async ({ userPreferences }) =>
    load<IR<SchemaLocalization>>(
      formatUrl('/context/schema_localization.json', {
        lang: userPreferences.get('form', 'schema', 'language'),
      }),
      'application/json'
    )
  )
);
export const getSchemaLocalization = (): IR<SchemaLocalization> =>
  schemaLocalization ??
  error('Accessing schema localization before fetching it');

const frontEndOnlyFields: Partial<Record<keyof Tables, RA<string>>> = {};
export const getFrontEndOnlyFields = (): Partial<
  RR<keyof Tables, RA<string>>
> => frontEndOnlyFields;

export const tables = {} as {
  readonly [TABLE_NAME in keyof Tables]: SpecifyTable<Tables[TABLE_NAME]>;
};
export const genericTables = tables as RR<keyof Tables, SpecifyTable>;

export const fetchContext = f
  .all({
    dataModel: load<RA<TableDefinition>>(
      '/context/datamodel.json',
      'application/json'
    ),
    localization: fetchSchemaLocalization(),
  })
  .then(({ dataModel, localization }) => {
    schemaLocalization = localization;
    dataModel
      .map((tableDefinition) => {
        const table = new SpecifyTable(tableDefinition);
        overwriteReadOnly(genericTables, table.name, table);
        return [tableDefinition, table] as const;
      })
      .forEach(([tableDefinition, table]) => {
        const [frontEndFields, callback] = (
          schemaExtras[table.name] as typeof schemaExtras['Agent'] | undefined
        )?.(table as SpecifyTable<Agent>) ?? [[]];
        const [literalFields, relationships] = split(
          frontEndFields.map((field) => {
            field.isReadOnly = true;
            field.overrides.isReadOnly = true;
            return field;
          }),
          (field) => field.isRelationship
        );

        overwriteReadOnly(
          table,
          'literalFields',
          processFields(
            tableDefinition.fields.map(
              (fieldDefinition) => new LiteralField(table, fieldDefinition)
            ),
            literalFields
          )
        );
        overwriteReadOnly(
          table,
          'relationships',
          processFields(
            tableDefinition.relationships.map(
              (relationshipDefinition) =>
                new Relationship(table, relationshipDefinition)
            ),
            relationships
          )
        );
        overwriteReadOnly(table, 'fields', [
          ...table.literalFields,
          ...table.relationships,
        ]);
        overwriteReadOnly(
          table,
          'field',
          Object.fromEntries(table.fields.map((field) => [field.name, field]))
        );

        frontEndOnlyFields[table.name] = [
          ...literalFields,
          ...relationships,
        ].map(({ name }) => name);

        callback?.();
      });
    return tables;
  });

setDevelopmentGlobal('_tables', genericTables);

/**
 * Returns a schema table object describing the named Specify model
 * Can wrap this function call in defined() to cast result to SpecifyTable
 *
 * @remarks
 * This function is useful when getting table using a dynamic name. If name is
 * known statically, can get table directly on the "tables" object
 */
export function getTable(name: string): SpecifyTable | undefined {
  if (
    process.env.NODE_ENV !== 'production' &&
    Object.keys(genericTables).length === 0
  )
    // eslint-disable-next-line functional/no-throw-statement
    throw new Error(
      `Calling getTable() before data model is fetched.${
        process.env.NODE_ENV === 'test'
          ? ' If this is part of a test, remember to include requireContext() at the top of the test file'
          : ''
      }`
    );

  const lowerCase = name.toLowerCase();
  return name === ''
    ? undefined
    : genericTables[name as keyof Tables] ??
        Object.values(genericTables).find(
          (table) => table.name.toLowerCase() === lowerCase
        ) ??
        Object.values(genericTables).find(
          (table) => table.longName.toLowerCase() === lowerCase
        );
}

export const strictGetTable = (name: string): SpecifyTable =>
  defined(getTable(name), `Trying to get unknown table: ${name}`);

export function getTreeTable(name: string): SpecifyTable<AnyTree> | undefined {
  const table = getTable(name);
  if (typeof table === 'object' && !isTreeTable(table.name))
    // eslint-disable-next-line functional/no-throw-statement
    throw new Error(`${name} is not a tree table`);
  return table as unknown as SpecifyTable<AnyTree> | undefined;
}

/**
 * Looks up a schema table object describing Specify model using the Specify
 * tableId integer
 */
export const getTableById = <SCHEMA extends AnySchema>(
  tableId: number
): SpecifyTable<SCHEMA> =>
  (Object.values(genericTables).find((table) => table.tableId === tableId) as
    | SpecifyTable<SCHEMA>
    | undefined) ?? error(`Table with id ${tableId} does not exist`);

// If this is true, then you can use {domainfilter:true} when fetching that table
export const hasHierarchyField = (table: SpecifyTable): boolean =>
  [
    'collectionObject',
    'collection',
    'discipline',
    'division',
    'institution',
  ].some((fieldName) =>
    table.relationships.some(({ name }) => name === fieldName)
  );
