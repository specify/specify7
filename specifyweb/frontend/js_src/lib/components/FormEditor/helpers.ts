import type { RR } from '../../utils/types';
import type { Tables } from '../DataModel/types';

export const tablesWithBusRulesIn6 = new Set([
  'AccessionAgentBusRules',
  'AccessionAuthorizationBusRules',
  'AccessionBusRules',
  'AddressBusRules',
  'AgentBusRules',
  'AppraisalBusRules',
  'AttachmentBusRules',
  'AuthorBusRules',
  'BioStratBusRules',
  'BorrowBusRules',
  'CatalogNumberingSchemeBusRules',
  'CatAutoNumberingSchemeBusRules',
  'CollectingEventBusRules',
  'CollectingTripAuthorizationBusRules',
  'CollectingTripBusRules',
  'CollectionBusRules',
  'CollectionObjectBusRules',
  'CollectorBusRules',
  'ConservDescriptionBusRules',
  'ConservEventBusRules',
  'ContainerBusRules',
  'DataTypeBusRules',
  'DeaccessionAgentBusRules',
  'DeterminationBusRules',
  'DeterminerBusRules',
  'DisciplineBusRules',
  'DivisionBusRules',
  'DNASequenceBusRules',
  'DNASequencingRunBusRules',
  'ExchangeOutBusRules',
  'ExchangeOutPrepBusRules',
  'ExtractorBusRules',
  'FieldNotebookBusRules',
  'FieldNotebookPageSetBusRules',
  'FundingAgentBusRules',
  'GeographyBusRules',
  'GeologicTimePeriodBusRules',
  'GiftBusRules',
  'GiftPreparationBusRules',
  'GroupPersonBusRules',
  'InfoRequestBusRules',
  'LithoStratBusRules',
  'LoanBusRules',
  'LoanGiftShipmentBusRules',
  'LoanPreparationBusRules',
  'LoanReturnPreparationBusRules',
  'LocalityBusRules',
  'PcrPersonBusRules',
  'PermitBusRules',
  'PickListBusRules',
  'PreparationBusRules',
  'PrepTypeBusRules',
  'SpecifyUserBusRules',
  'StorageBusRules',
  'TaxonBusRules',
  'TreatmentEventBusRules',
  'TreeDefBusRules',
]);

/**
 * Most of the time business rules class name can be inferred from table name.
 * Exceptions:
 */
export const businessRulesOverride: Partial<RR<keyof Tables, string>> = {
  Shipment: 'LoanGiftShipment',
  CollectionObjectAttachment: 'Attachment',
};

export const getBusinessRuleClassFromTable = (
  tableName: keyof Tables
): string =>
  businessRulesOverride[tableName] === undefined
    ? tablesWithBusRulesIn6.has(`${tableName}BusRules`)
      ? `edu.ku.brc.specify.datamodel.busrules.${tableName}BusRules`
      : ''
    : `edu.ku.brc.specify.datamodel.busrules.${businessRulesOverride[
        tableName
      ]!}BusRules`;
