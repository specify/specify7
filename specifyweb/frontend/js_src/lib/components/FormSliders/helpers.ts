import { Relationship } from '../DataModel/specifyField';

export const shouldBeToOne = (
  relationship: Relationship | undefined
): boolean =>
  relationship?.type.includes('-to-many') === false ||
  (relationship?.type.includes('-to-many') === true &&
    !relationship.isDependent() &&
    relationship.getReverse()?.isDependent() === true);
