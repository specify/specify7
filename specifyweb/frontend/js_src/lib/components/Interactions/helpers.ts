import { ajax } from '../../utils/ajax';
import { formData } from '../../utils/ajax/helpers';
import type { RA, RestrictedTuple } from '../../utils/types';
import type { AnyInteractionPreparation } from '../DataModel/helperTypes';
import type { Tables } from '../DataModel/types';

export const interactionPrepTables: RestrictedTuple<
  AnyInteractionPreparation['tableName']
> = [
  'DisposalPreparation',
  'ExchangeInPrep',
  'ExchangeOutPrep',
  'GiftPreparation',
  'LoanPreparation',
];

type ExtractInteraction<T extends string> =
  T extends `${infer Prefix}Prep${string}` ? Prefix : never;

export type InteractionWithPreps = Tables[ExtractInteraction<
  AnyInteractionPreparation['tableName']
>];

export const interactionsWithPrepTables: RestrictedTuple<
  InteractionWithPreps['tableName']
> = ['Disposal', 'ExchangeIn', 'ExchangeOut', 'Gift', 'Loan'];

export type PreparationData = {
  readonly catalogNumber: string;
  readonly collectionObjectId: number;
  readonly taxon: string;
  readonly taxonId: number;
  readonly preparationId: number;
  readonly prepType: string;
  readonly countAmount: number;
  readonly loaned: number;
  readonly gifted: number;
  readonly exchanged: number;
  readonly available: number;
};

export type PreparationRow = readonly [
  catalogNumber: string,
  collectionObjectId: number,
  taxonFullName: string,
  taxonId: number,
  preparationId: number,
  prepType: string,
  preparationCountAmt: number,
  amountLoaned: string | null,
  amountedGifted: string | null,
  amountExchanged: string | null,
  amountAvailable: string
];

export const getPrepsAvailableForLoanRs = async (
  recordSetId: number,
  isLoan: boolean
) =>
  ajax<RA<PreparationRow>>(
    `/interactions/preparations_available_rs/${recordSetId}/`,
    {
      method: 'POST',
      headers: { Accept: 'application/json' },
      body: formData({ isLoan }),
    }
  ).then(({ data }) => data);

export const getPrepsAvailableForLoanCoIds = async (
  idField: string,
  collectionObjectIds: RA<string>,
  isLoan: boolean
) =>
  ajax<RA<PreparationRow>>('/interactions/preparations_available_ids/', {
    method: 'POST',
    headers: { Accept: 'application/json' },
    body: formData({
      id_fld: idField,
      co_ids: collectionObjectIds,
      isLoan,
    }),
  }).then(({ data }) => data);

export const getInteractionsForPrepId = async (prepId: number) =>
  ajax<RA<readonly [number, string | null, string | null, string | null]>>(
    '/interactions/prep_interactions/',
    {
      method: 'POST',
      headers: { Accept: 'application/json' },
      body: formData({ prepIds: prepId }),
    }
  ).then(({ data }) => data[0]);

export const getPrepAvailability = async (
  prepId: number,
  iPrepId: number | undefined,
  iPrepName: string
) =>
  ajax(
    `/interactions/prep_availability/${prepId}/${
      typeof iPrepId === 'number' ? `${iPrepId}/${iPrepName}/` : ''
    }`,
    { headers: { Accept: 'application/json' } }
  ).then(({ data }) => data);
