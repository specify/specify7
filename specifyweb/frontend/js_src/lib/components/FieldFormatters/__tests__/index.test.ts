import { mockTime, requireContext } from '../../../tests/helpers';
import { getField } from '../../DataModel/helpers';
import { tables } from '../../DataModel/tables';
import type { UiFormatter } from '../index';
import { fetchContext, getUiFormatters } from '../index';

mockTime();
requireContext();

test('field formatters are fetched and parsed correctly', async () =>
  expect(fetchContext).resolves.toMatchSnapshot());

const getFormatter = (): UiFormatter | undefined =>
  getField(tables.CollectionObject, 'catalogNumber')?.getUiFormatter();

const getSecondFormatter = (): UiFormatter | undefined =>
  getUiFormatters().AccessionNumber;

describe('valueOrWild', () => {
  test('catalog number', () =>
    expect(getFormatter()?.valueOrWild()).toBe('#########'));

  test('accession number', () =>
    expect(getSecondFormatter()?.valueOrWild()).toBe('2022-AA-###'));
});

describe('parseRegExp', () => {
  test('catalog number', () =>
    expect(getFormatter()?.parseRegExp()).toBe('^(#########|\\d{0,9})$'));
  test('accession number', () =>
    expect(getSecondFormatter()?.parseRegExp()).toBe(
      '^(YEAR|\\d{4})(-)([a-zA-Z0-9]{2})(-)(###|\\d{3})$'
    ));
});

describe('pattern', () => {
  test('catalog number', () =>
    expect(getUiFormatters().CatalogNumberNumericRegex?.pattern()).toBe(
      '####[-A]'
    ));
});

describe('format', () => {
  test('catalog number', () =>
    expect(getFormatter()?.format('4')).toBe('000000004'));

  test('overly long catalog number', () =>
    expect(getFormatter()?.format('0000000004')).toBeUndefined());

  test('invalid catalog number', () =>
    expect(getFormatter()?.format('a')).toBeUndefined());
});
