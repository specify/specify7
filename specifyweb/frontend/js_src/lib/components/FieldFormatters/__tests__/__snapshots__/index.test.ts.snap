// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field formatters are fetched and parsed correctly 1`] = `
{
  "AccessionNumber": UiFormatter {
    "fields": [
      YearField {
        "autoIncrement": false,
        "byYear": true,
        "pattern": undefined,
        "size": 4,
        "value": "YEAR",
      },
      SeparatorField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 1,
        "value": "-",
      },
      AlphaNumberField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 2,
        "value": "AA",
      },
      SeparatorField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 1,
        "value": "-",
      },
      NumericField {
        "autoIncrement": true,
        "byYear": false,
        "pattern": undefined,
        "size": 3,
        "value": "###",
      },
    ],
    "isSystem": true,
    "table": "[table Accession]",
    "title": "AccessionNumber",
  },
  "AccessionNumberByYear": UiFormatter {
    "fields": [
      YearField {
        "autoIncrement": false,
        "byYear": true,
        "pattern": undefined,
        "size": 4,
        "value": "YEAR",
      },
      SeparatorField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 1,
        "value": "-",
      },
      AlphaNumberField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 2,
        "value": "AA",
      },
      SeparatorField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 1,
        "value": "-",
      },
      AlphaNumberField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 3,
        "value": "AAA",
      },
    ],
    "isSystem": true,
    "table": "[table Accession]",
    "title": "AccessionNumberByYear",
  },
  "AccessionStringFormatter": UiFormatter {
    "fields": [
      AlphaNumberField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 10,
        "value": "AAAAAAAAAA",
      },
    ],
    "isSystem": true,
    "table": "[table Accession]",
    "title": "AccessionStringFormatter",
  },
  "CatalogNumber": UiFormatter {
    "fields": [
      YearField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 4,
        "value": "YEAR",
      },
      SeparatorField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 1,
        "value": "-",
      },
      NumericField {
        "autoIncrement": true,
        "byYear": false,
        "pattern": undefined,
        "size": 6,
        "value": "######",
      },
    ],
    "isSystem": false,
    "table": "[table CollectionObject]",
    "title": "CatalogNumber",
  },
  "CatalogNumberAlphaNumByYear": UiFormatter {
    "fields": [
      YearField {
        "autoIncrement": false,
        "byYear": true,
        "pattern": undefined,
        "size": 4,
        "value": "YEAR",
      },
      SeparatorField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 1,
        "value": "-",
      },
      NumericField {
        "autoIncrement": true,
        "byYear": false,
        "pattern": undefined,
        "size": 6,
        "value": "######",
      },
    ],
    "isSystem": false,
    "table": "[table CollectionObject]",
    "title": "CatalogNumberAlphaNumByYear",
  },
  "CatalogNumberNumeric": CatalogNumberNumeric {
    "fields": [
      CatalogNumberNumericField {
        "autoIncrement": true,
        "byYear": false,
        "pattern": undefined,
        "size": 9,
        "value": "#########",
      },
    ],
    "isSystem": true,
    "table": "[table CollectionObject]",
    "title": "Catalog Number Numeric",
  },
  "CatalogNumberNumericRegex": UiFormatter {
    "fields": [
      RegexField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": "####[-A]",
        "size": 1,
        "value": "[0-9]{4}(-[A-Z])?",
      },
    ],
    "isSystem": false,
    "table": "[table CollectionObject]",
    "title": "CatalogNumberNumericRegex",
  },
  "Date": UiFormatter {
    "fields": [],
    "isSystem": true,
    "table": undefined,
    "title": "Date",
  },
  "DeaccessionNumber": UiFormatter {
    "fields": [
      YearField {
        "autoIncrement": false,
        "byYear": true,
        "pattern": undefined,
        "size": 4,
        "value": "YEAR",
      },
      SeparatorField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 1,
        "value": "-",
      },
      AlphaNumberField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 2,
        "value": "AA",
      },
      SeparatorField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 1,
        "value": "-",
      },
      NumericField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 3,
        "value": "###",
      },
    ],
    "isSystem": true,
    "table": "[table Deaccession]",
    "title": "DeaccessionNumber",
  },
  "GiftNumber": UiFormatter {
    "fields": [
      YearField {
        "autoIncrement": false,
        "byYear": true,
        "pattern": undefined,
        "size": 4,
        "value": "YEAR",
      },
      SeparatorField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 1,
        "value": "-",
      },
      NumericField {
        "autoIncrement": true,
        "byYear": false,
        "pattern": undefined,
        "size": 3,
        "value": "###",
      },
    ],
    "isSystem": true,
    "table": "[table Gift]",
    "title": "GiftNumber",
  },
  "InfoRequestNumber": UiFormatter {
    "fields": [
      YearField {
        "autoIncrement": false,
        "byYear": true,
        "pattern": undefined,
        "size": 4,
        "value": "YEAR",
      },
      SeparatorField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 1,
        "value": "-",
      },
      NumericField {
        "autoIncrement": true,
        "byYear": false,
        "pattern": undefined,
        "size": 3,
        "value": "###",
      },
    ],
    "isSystem": true,
    "table": "[table InfoRequest]",
    "title": "InfoRequestNumber",
  },
  "KUITeach": CatalogNumberNumeric {
    "fields": [
      CatalogNumberNumericField {
        "autoIncrement": true,
        "byYear": false,
        "pattern": undefined,
        "size": 9,
        "value": "#########",
      },
    ],
    "isSystem": true,
    "table": "[table CollectionObject]",
    "title": "Catalog Number Numeric",
  },
  "LoanNumber": UiFormatter {
    "fields": [
      YearField {
        "autoIncrement": false,
        "byYear": true,
        "pattern": undefined,
        "size": 4,
        "value": "YEAR",
      },
      SeparatorField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 1,
        "value": "-",
      },
      NumericField {
        "autoIncrement": true,
        "byYear": false,
        "pattern": undefined,
        "size": 3,
        "value": "###",
      },
    ],
    "isSystem": true,
    "table": "[table Loan]",
    "title": "LoanNumber",
  },
  "NumericBigDecimal": UiFormatter {
    "fields": [
      NumericField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 15,
        "value": "###############",
      },
    ],
    "isSystem": true,
    "table": undefined,
    "title": "NumericBigDecimal",
  },
  "NumericByte": UiFormatter {
    "fields": [
      NumericField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 3,
        "value": "###",
      },
    ],
    "isSystem": true,
    "table": undefined,
    "title": "NumericByte",
  },
  "NumericDouble": UiFormatter {
    "fields": [
      NumericField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 10,
        "value": "##########",
      },
    ],
    "isSystem": true,
    "table": undefined,
    "title": "NumericDouble",
  },
  "NumericFloat": UiFormatter {
    "fields": [
      NumericField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 10,
        "value": "##########",
      },
    ],
    "isSystem": true,
    "table": undefined,
    "title": "NumericFloat",
  },
  "NumericInteger": UiFormatter {
    "fields": [
      NumericField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 10,
        "value": "##########",
      },
    ],
    "isSystem": true,
    "table": undefined,
    "title": "NumericInteger",
  },
  "NumericLong": UiFormatter {
    "fields": [
      NumericField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 10,
        "value": "##########",
      },
    ],
    "isSystem": true,
    "table": undefined,
    "title": "NumericLong",
  },
  "NumericShort": UiFormatter {
    "fields": [
      NumericField {
        "autoIncrement": false,
        "byYear": false,
        "pattern": undefined,
        "size": 5,
        "value": "#####",
      },
    ],
    "isSystem": true,
    "table": undefined,
    "title": "NumericShort",
  },
  "PartialDate": UiFormatter {
    "fields": [],
    "isSystem": false,
    "table": undefined,
    "title": "PartialDate",
  },
  "PartialDateMonth": UiFormatter {
    "fields": [],
    "isSystem": false,
    "table": undefined,
    "title": "PartialDateMonth",
  },
  "PartialDateYear": UiFormatter {
    "fields": [],
    "isSystem": false,
    "table": undefined,
    "title": "PartialDateYear",
  },
  "SearchDate": UiFormatter {
    "fields": [],
    "isSystem": true,
    "table": undefined,
    "title": "SearchDate",
  },
}
`;
